spec/support/shared_examples/block_matcher.rb

```diff
@@ -0,0 +1,102 @@
+RSpec.shared_examples "an RSpec block-only matcher" do
+  # Note: do not use `matcher` in 2 expectation expressions in a single
+  # example here. In some cases (such as `change { }.to(2)`), it will fail
+  # because using it a second time will apply `x += 2` twice, changing
+  # the value to 4.
+
+  matcher :always_passes do
+    supports_block_expectations
+    match do |actual|
+      actual.call
+      true
+    end
+  end
+
+  matcher :always_fails do
+    supports_block_expectations
+    match do |actual|
+      actual.call
+      false
+    end
+  end
+
+  def valid_expectation
+    expect { |block| valid_block(&block) }
+  end
+
+  def invalid_expectation
+    expect { |block| invalid_block(&block) }
+  end
+
+  include_examples "an RSpec matcher"
+
+  it 'preserves the symmetric property of `==`' do
+    expect(matcher).to eq(matcher)
+    # TODO: the following expectations fail due to
+    #
+    # 1) output.to_stderr matcher behaves like an RSpec block-only matcher preserves the symmetric property of `==`
+    #    Failure/Error: raise "Warnings were generated: #{output}" if has_output?
+    #    RuntimeError:
+    #      Warnings were generated: foo
```

Comment:
So this is because warnings are generated by the specs, it should be possible to rewrite the relevant specs warning free...

Reply:
Previously, valid values were defined as e.g.:
```ruby
  it_behaves_like "an RSpec matcher", :valid_value => lambda { k += 1 },
```
and that were supposed to check that
```ruby
expect { k += 1 }.to change { k }
```
The `matcher` here is initialized by `change { k }`, and the target is a block.

So the whole point of those two expectations is to make sure that `==` on both the lambda and the matcher return `false` when compared with each other.

After this change the target block is defined using a method, and the reason for warnings to appear is that those `output`/`throw_symbol`/`raise_error` targets are actually get called when target is compared, since they are not a block.

Do you think that writing those examples as the following will achieve the same goal?
```ruby
    expect(matcher).not_to eq(-> { valid_block })
    expect(-> { valid_block }).not_to eq(matcher)
```